### PR TITLE
feat(twig): function users by role and circles

### PR DESCRIPTION
**Pull request by IsaMic**
|Q              |A  |
|---------------|---|
|Bug fix?       |n|
|New feature?   |y|	
|BC breaks?     |n|	
|Deprecations?  |n|	
|Fixed tickets  |n|	

twig function to retrieve users for a role with circles.

### CODE REVIEW

File: src/Twig/AppExtension.php [hunk 1]
```diff
@@ -130,6 +130,7 @@ public function getFunctions()
             new TwigFunction('call_user_func', array($this, 'callUserFunc')),
             new TwigFunction('emsco_generate_email', array($this, 'generateEmailMessage')),
             new TwigFunction('emsco_send_email', array($this, 'sendEmail')),
+            new TwigFunction('emsco_users_by_role_and_circles', [UserService::class, 'getUsersForRoleAndCircles']),
```
Comment (dameert) [inline src/Twig/AppExtension.php:133]: This is exposing a DB query directly from the repository directly in TWIG.
This is also very specific.

You could create a dedicated runtime UserRuntime that creates a UserList object. This object should have the functionality to filter based on role and/or circles. This is much more flexible for future userList features. 

Possible feature request: only show the enabled users (from the full UserList)

**Review by dameert — CHANGES_REQUESTED** [review]

File: src/Twig/AppExtension.php (hunk 1 above)
Comment (Davidmattei) [reply src/Twig/AppExtension.php:133]: 👍

**Review by theus77 — DISMISSED** [review]
@IsaMic tu peux merger

File: src/Twig/UserRuntime.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    private $userRepository;
+    
+    public function __construct(UserRepository $userRepository)
+    {
+        $this->userRepository = $userRepository;
+    }
+    
+    /**
+     * @return UserList
+     */
```
Comment (dameert) [inline src/Twig/UserRuntime.php:24]:
```suggestion
```

File: src/Twig/UserRuntime.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+declare(strict_types=1);
+
+namespace EMS\CoreBundle\Twig;
+
+use EMS\CoreBundle\Core\User\UserList;
+use EMS\CoreBundle\Repository\UserRepository;
+use Twig\Extension\RuntimeExtensionInterface;
+
+
+class UserRuntime implements RuntimeExtensionInterface
```
Comment (dameert) [inline src/Twig/UserRuntime.php:12]:
```suggestion
final class UserRuntime implements RuntimeExtensionInterface
```

File: src/Repository/UserRepositoryInterface.php [hunk 4]
```diff
@@ -10,5 +10,17 @@
...
+     * @param array<string> $circles
+     * @return array <\EMS\CoreBundle\Entity\User>
+     */
     public function findForRoleAndCircles($role, $circles);
+    
+    /**
+     *
+     * @return \EMS\CoreBundle\Core\User\UserList
+     */
+    public function getUsersEnabled();
```
Comment (dameert) [inline src/Repository/UserRepositoryInterface.php:25]:
```suggestion
    /**
     *
     * @return \EMS\CoreBundle\Core\User\UserList
     */
    public function getUsersEnabled(): UserList;
```

File: src/Repository/UserRepository.php [hunk 5]
```diff
@@ -11,6 +13,9 @@
  */
 class UserRepository extends \Doctrine\ORM\EntityRepository implements UserRepositoryInterface
 {
+    /**
+     * {@inheritDoc}
+     */
```
Comment (dameert) [inline src/Repository/UserRepository.php:18]:
```suggestion
```

File: src/Repository/UserRepository.php [hunk 6]
```diff
@@ -31,4 +36,15 @@ public function findForRoleAndCircles($role, $circles)
         }
         return $resultSet;
     }
+    
+    /**
+     *  {@inheritDoc}
+     */
+    public function getUsersEnabled() : UserList
```
Comment (dameert) [inline src/Repository/UserRepository.php:43]:
```suggestion

    public function getUsersEnabled(): UserList
```

File: src/Core/User/UserList.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+     * @param array<User> $users
+     */
+    public function __construct(array $users)
+    {
+        $this->users = $users;
+    }
+    
+    /**
+     *
+     * @return array <\EMS\CoreBundle\Entity\User>
```
Comment (dameert) [inline src/Core/User/UserList.php:22]:
```suggestion
     * @return array <User>
```

File: src/Core/User/UserList.php [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    public function getUsers(): array
+    {
+        return $this->users;
+    }
+    /**
+     * @param array<string> $roles
+     * @return UserList
```
Comment (dameert) [inline src/Core/User/UserList.php:30]:
```suggestion
```

File: src/Core/User/UserList.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    public function getForRoles(array $roles): UserList
+    {
+        $usersList = [];
+        /** \EMS\CoreBundle\Entity\User $user */
```
Comment (dameert) [inline src/Core/User/UserList.php:35]:
```suggestion
        /** User $user */
```

**Review by dameert — CHANGES_REQUESTED** [review]
small changes. good work

**Review by dameert — APPROVED** [review]

**Pull request merged**